File: examples/Python.ExampleApplication/Hooks/transformers/action_url_xml_transformer.py
from xml.etree import ElementTree
from tableau_migration import (
    IPublishableWorkbook,
    XmlContentTransformerBase
)

class ActionUrlXmlTransformer(XmlContentTransformerBase[IPublishableWorkbook]):
    
    def needs_xml_transforming(self, ctx: IPublishableWorkbook) -> bool:
        # Returning false prevents the transform method from running.
        # Implementing this method potentially allows workbooks to migrate
        # without loading the file into memory, improving migration speed.
        return True

    def transform(self, ctx: IPublishableWorkbook, xml: ElementTree.Element) -> None:
        # Changes to the XML are saved back to the workbook file before publishing.
        for action_link in xml.findall("actions/*/link"):
            action_link.set("expression", action_link.get("expression").replace("127.0.0.1", "testserver"))

File: examples/Csharp.ExampleApplication/Hooks/Transformers/ActionUrlXmlTransformer.cs
using System.Threading;
using System.Threading.Tasks;
using System.Xml.Linq;
using System.Xml.XPath;
using Tableau.Migration.Content;
using Tableau.Migration.Engine.Hooks.Transformers;

namespace Csharp.ExampleApplication.Hooks.Transformers
{
    #region class

    public class ActionUrlXmlTransformer : XmlContentTransformerBase<IPublishableWorkbook>
    {
        protected override bool NeedsXmlTransforming(IPublishableWorkbook ctx)
        {
            /*
             * Returning false prevents TransformAsync from running.
             * Implementing this method potentially allows workbooks to migrate without 
             * loading the file into memory, improving migration speed.
             */
            return true;
        }

        public override Task TransformAsync(IPublishableWorkbook ctx, XDocument xml, CancellationToken cancel)
        {
            // Changes to the XML are saved back to the workbook file before publishing.
            foreach (var actionLink in xml.XPathSelectElements("//actions/*/link"))
            {
                actionLink.SetAttributeValue("expression", actionLink.Attribute("expression")?.Value?.Replace("127.0.0.1", "testserver"));
            }

            return Task.CompletedTask;
        }
    }

    #endregion
}
